ocaml/typing/typecore.ml

```diff
@@ -6992,49 +6997,55 @@ and type_cases
-                type_match
-                  (* Pattern guards containing no value cases will have an
-                     "Any" [_] case inserted during translation to handle the
-                     case where no cases match, so we can successfully
-                     typecheck such guards. Accordingly, [require_value_case]
-                     is set to [false] below. *)
-                  ~require_value_case:false ppg_scrutinee ppg_cases ext_env
-                  ppg_loc Check_and_warn_if_total
-                  (mk_expected ?explanation ty_res') emode
-              in
-              Pattern_guarded_rhs
-                { pg_scrutinee = arg
-                ; pg_scrutinee_sort = sort
-                ; pg_cases = cases
-                ; pg_partial = partial
-                ; pg_loc = ppg_loc
-                ; pg_env = ext_env
-                ; pg_type = instance ty_res'
-                }
+          match Jane_syntax.Case.of_ast untyped_case with
+          | Some (jcase, _attrs) ->
+              (match jcase with
+               | Jcase_pattern_guarded (Pg_case { scrutinee; cases }) ->
+                   let loc = untyped_case.pc_rhs.pexp_loc in
+                   let { arg; sort; cases; partial; } =
+                     type_match
+                       (* Pattern guards containing no value cases will have an
+                          "Any" [_] case inserted during translation to handle
+                          the case where no cases match, so we can successfully
+                          typecheck such guards. Accordingly,
+                          [require_value_case] is set to [false] below. *)
+                       ~require_value_case:false scrutinee cases ext_env loc
+                       Check_and_warn_if_total
+                       (mk_expected ?explanation ty_res') emode
+                   in
+                   Pattern_guarded_rhs
+                     { pg_scrutinee = arg
+                     ; pg_scrutinee_sort = sort
+                     ; pg_cases = cases
+                     ; pg_partial = partial
+                     ; pg_loc = loc
+                     ; pg_env = ext_env
+                     ; pg_type = instance ty_res'
+                     })
+          | None ->
+          let typed_guard =
+            match untyped_case.pc_guard with
+            | None -> None
+            | Some guard ->
+                Some
+                  (type_expect ext_env mode_local guard
+                     (mk_expected ~explanation:When_guard Predef.type_bool))
+          in
+          let typed_rhs =
+            { (type_expect ?in_function ext_env emode untyped_case.pc_rhs
+                 (mk_expected ?explanation ty_res'))
+              with exp_type = instance ty_res' }
+          in
+          match typed_guard with
+          | None -> Simple_rhs typed_rhs
+          | Some typed_guard ->
+              Boolean_guarded_rhs { bg_guard = typed_guard; bg_rhs = typed_rhs }
```

Comment:
I was somewhat surprised at how large this diff is. (I mean, I think I see that the diff is because of reordering/indentation rather than bona fide code changes.) It would be nice to pare it down a bit more, if that's possible. (It might not be.)

Reply:
Resolved by the above